pkg/agent/controller/networkpolicy/networkpolicy_controller_test.go

```diff
@@ -507,6 +515,176 @@ func TestAddNetworkPolicyWithMultipleRules(t *testing.T) {
 	assert.Equal(t, 1, controller.GetAppliedToGroupNum())
 }
 
+func writeToFile(t *testing.T, fs afero.Fs, dir, file string, base64Str string) {
+	data, err := base64.StdEncoding.DecodeString(base64Str)
+	require.NoError(t, err)
+	f, err := fs.OpenFile(dir+"/"+file, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o600)
+	require.NoError(t, err)
+	defer f.Close()
+	_, err = f.Write(data)
+	require.NoError(t, err)
+}
+
+func TestFallbackToFileStore(t *testing.T) {
+	prepareMockTables()
+	tests := []struct {
+		name          string
+		initFileStore func(networkPolicyStore, appliedToGroupStore, addressGroupStore *fileStore)
+		expectedRule  *CompletedRule
+	}{
+		{
+			name: "same storage version",
+			initFileStore: func(networkPolicyStore, appliedToGroupStore, addressGroupStore *fileStore) {
+				networkPolicyStore.save(newNetworkPolicy("policy1", "uid1", []string{"addressGroup1"}, nil, []string{"appliedToGroup1"}, nil))
+				appliedToGroupStore.save(newAppliedToGroup("appliedToGroup1", []v1beta2.GroupMember{*newAppliedToGroupMemberPod("pod1", "namespace")}))
+				addressGroupStore.save(newAddressGroup("addressGroup1", []v1beta2.GroupMember{*newAddressGroupPodMember("pod2", "namespace", "192.168.0.1")}))
+			},
+			expectedRule: &CompletedRule{
+				rule: &rule{
+					Direction:       v1beta2.DirectionIn,
+					From:            v1beta2.NetworkPolicyPeer{AddressGroups: []string{"addressGroup1"}},
+					MaxPriority:     -1,
+					AppliedToGroups: []string{"appliedToGroup1"},
+					PolicyUID:       "uid1",
+					PolicyName:      "uid1",
+					SourceRef: &v1beta2.NetworkPolicyReference{
+						Type:      v1beta2.K8sNetworkPolicy,
+						Namespace: testNamespace,
+						Name:      "policy1",
+						UID:       "uid1",
+					},
+				},
+				FromAddresses: v1beta2.NewGroupMemberSet(newAddressGroupPodMember("pod2", "namespace", "192.168.0.1")),
+				TargetMembers: v1beta2.NewGroupMemberSet(newAppliedToGroupMemberPod("pod1", "namespace")),
+			},
+		},
+		{
+			// The test is to ensure compatibility with v1beta2 storage version if one day the used version is upgraded.
+			name: "compatible with v1beta2",
+			initFileStore: func(networkPolicyStore, appliedToGroupStore, addressGroupStore *fileStore) {
+				// The bytes of v1beta2 objects serialized in protobuf.
+				// They are not supposed to be updated when bumping up the used version.
+				base64EncodedPolicy := "azhzAAovCh5jb250cm9scGxhbmUuYW50cmVhLmlvL3YxYmV0YTISDU5ldHdvcmtQb2xpY3kSdAoYCgR1aWQxEgAaACIAKgR1aWQxMgA4AEIAEh8KAkluEg8KDWFkZHJlc3NHcm91cDEaACgAOABKAFoAGg9hcHBsaWVkVG9Hcm91cDEyJgoQSzhzTmV0d29ya1BvbGljeRIDbnMxGgdwb2xpY3kxIgR1aWQxGgAiAA=="
```

Comment:
I am not a big fan of hardcoding the test strings like this. Would it be possible to generate the encoding on the fly from an actual v1beta2 object?

Reply:
It's made so because when agent switches to next version in the future, all v1beta2 reference will likely be updated to the next version directly, then the code supposed to generate v1beta2 object will generate an object of new version, which makes the test not different from "same storage version" test.